python/metatensor-learn/metatensor/learn/nn/equivariant_transform.py

```diff
@@ -0,0 +1,131 @@
+from typing import List, Optional, Union
+
+import torch
+from torch.nn import Module
+
+from .._backend import Labels, TensorMap
+from .module_map import ModuleMap
+from ._utils import _check_module_map_parameter
+
+
+class EquivariantTransform(Module):
+    """
+    Applies a module to invariant blocks and to the invariants commputed from each
+    covariant block of a :py:class:`TensorMap` passed to its forward method, indexed by
+    :param in_keys:. For covariant blocks, the transformed values are then elementwise
+    multiplied to the original block, without affecting its `component` dimension.
+
+    Each parameter can be passed as a single value of its expected type, which is used
+    as the parameter for all blocks. Alternatively, they can be passed as a list to
+    control the parameters applied to each block indexed by the keys in :param in_keys:.
+
+    :param module: :py:class:`callable`, the callable to apply to the invariant blocks
+        and to the invariant built from the covariant blocks.
+    :param in_keys: :py:class:`Labels`, the keys that are assumed to be in the input
+        tensor map in the :py:meth:`forward` method.
+    :param invariant_key_idxs: list of int, the indices of the invariant keys present in
+        `in_keys` in the input :py:class:`TensorMap`. Only blocks for these keys will
+        have bias applied according to the user choice. Covariant blocks will not have
+        bias applied.
+    :param in_features: list of int, the number of features in the input tensor for each
+        block indexed by the keys in :param in_keys:. If passed as a single value, the
+        same number of features is assumed for all blocks.
+    :param out_properties: list of :py:class`Labels` (optional), the properties labels
+        of the output. By default the output properties are relabeled using
+        Labels.range. If provided, :param out_features: can be inferred and need not be
+        provided.
+    :param device: :py:class:`torch.device` Device to instantiate the module in.
+    :param dtype: :py:class:`torch.dtype` passed to the module instance.
+    """
+
+    def __init__(
+        self,
+        module: callable,
+        in_keys: Labels,
+        invariant_key_idxs: List[int],
+        in_features: Union[int, List[int]],
+        out_features: Optional[Union[int, List[int]]] = None,
```

Comment:
`out_features` are not needed here as they are the same as `in_features`. The  user can still pass `out_properties` if they want the output TensorMap to have the correct properties metadata, i.e. passing:
```py
EquivariantTransform(
    ...,
    out_properties=[tensor[key].block.properties for key in in_keys],
    ...,
)
```

```suggestion
```